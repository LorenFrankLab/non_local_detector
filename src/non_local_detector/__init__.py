from non_local_detector.models import (
    ContFragClusterlessClassifier,  # noqa
    ContFragSortedSpikesClassifier,  # noqa
    MultiEnvironmentSortedSpikesClassifier,  # noqa
    MultiEnvironmentClusterlessClassifier,  # noqa
    NonLocalClusterlessDetector,  # noqa
    NonLocalSortedSpikesDetector,  # noqa
    ClusterlessDecoder,  # noqa
    SortedSpikesDecoder,  # noqa
)

try:
    from ._version import __version__
except ImportError:
    pass
