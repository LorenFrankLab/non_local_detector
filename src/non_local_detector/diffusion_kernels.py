from typing import Any

import jax
import jax.numpy as jnp
import jax.scipy.linalg
import networkx as nx
import numpy as np


def _assign_gaussian_weights_from_distance(G: nx.Graph, bandwidth_sigma: float) -> None:
    """
    Overwrites each edge's "weight" attribute with
        w_uv = exp( - (distance_uv)^2 / (2 * sigma^2) ).
    Assumes each edge already has "distance" = Euclidean length.
    """
    two_sigma2 = 2.0 * (bandwidth_sigma**2)
    for u, v, data in G.edges(data=True):
        d = data.get("distance", None)
        if d is None:
            raise KeyError(f"Edge ({u},{v}) has no 'distance' attribute.")
        # compute the RBF weight
        data["weight"] = float(np.exp(-(d * d) / two_sigma2))


def compute_diffusion_kernels(
    graph: nx.Graph,
    bandwidth_sigma: float,
) -> jnp.ndarray:
    """
    Computes diffusion kernels for all valid bins using the
    Graph Laplacian method and matrix exponential.

    Parameters
    ----------
    graph : nx.Graph
        The graph representing the track, where nodes are bins and edges are connections.
    bandwidth_sigma : float
        The bandwidth of the Gaussian kernel. This controls the spread of the kernel.

    Returns
    -------
    kernel_matrix : jnp.ndarray, shape (n_bins, n_bins)
        The diffusion kernel matrix for the bins.
        Therepresents the amount of concentration (or probability mass, if u represents probability)
        that has flowed from node j to node i after time t.
    """

    n_bins = graph.number_of_nodes()
    _assign_gaussian_weights_from_distance(graph, bandwidth_sigma)
    laplacian_full = jnp.array(
        nx.laplacian_matrix(graph, nodelist=range(n_bins), weight="weight").toarray(),
        dtype=jnp.float32,
    )

    exponent_coefficient = bandwidth_sigma**2 / 2.0
    kernel_matrix = jax.scipy.linalg.expm(-exponent_coefficient * laplacian_full)

    # Clip to non-negative values
    kernel_matrix = jnp.clip(kernel_matrix, a_min=0.0, a_max=None)

    # Normalize the kernel matrix so the sum of each column is 1.
    col_sums = kernel_matrix.sum(axis=0, keepdims=True)
    kernel_matrix = jnp.where(col_sums == 0.0, 0.0, kernel_matrix / col_sums)

    return kernel_matrix
