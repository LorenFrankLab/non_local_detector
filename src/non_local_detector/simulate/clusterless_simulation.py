"""Simulate clusterless spikes and associated spike waveform features.

Simulator return contract (decoder-ready):
- spike_times: List[np.ndarray], per electrode, seconds, strictly increasing
- spike_waveform_features: List[np.ndarray], per electrode, shape (n_spikes_e, n_features)
- edges: decoding bin edges in seconds, shape (n_time_bins+1), bin_widths = np.diff(edges)
- position_time / position: used for local decoding & encoding interpolation
All times are in SECONDS. No NaNs in marks. Empty electrodes: (0,), (0, n_features).

See tests._sim_contract.ClusterlessSimOutput for the full contract definition.
"""

import numpy as np

from non_local_detector.environment import Environment
from non_local_detector.simulate.simulate import (
    get_trajectory_direction,
    simulate_multiunit_with_place_fields,
    simulate_position,
    simulate_time,
)
from non_local_detector.tests._sim_contract import ClusterlessSimOutput

SAMPLING_FREQUENCY = 1000
TRACK_HEIGHT = 175
RUNNING_SPEED = 15
PLACE_FIELD_VARIANCE = 12.5
PLACE_FIELD_MEANS = np.arange(0, 200, 10)
N_RUNS = 15
REPLAY_SPEEDUP = 120.0
N_TETRODES = 5
N_FEATURES = 4
MARK_SPACING = 5


def make_simulated_run_data(
    sampling_frequency: int = SAMPLING_FREQUENCY,
    track_height: float = TRACK_HEIGHT,
    running_speed: float = RUNNING_SPEED,
    n_runs: int = N_RUNS,
    place_field_variance: float = PLACE_FIELD_VARIANCE,
    place_field_means: np.ndarray = PLACE_FIELD_MEANS,
    n_tetrodes: int = N_TETRODES,
    make_inbound_outbound_neurons: bool = False,
    seed: int | None = 0,
) -> ClusterlessSimOutput:
    """Make simulated data of a rat running back and forth on a linear maze.

    Generates clusterless spike data (spike times + waveform features) for a
    simulated animal running on a linear track. Returns decoder-ready outputs
    with per-electrode spike lists (no NaN padding).

    Parameters
    ----------
    sampling_frequency : int, optional
        Samples per second for position sampling. Default is 1000 Hz.
    track_height : float, optional
        Height of the simulated track in spatial units. Default is 175.
    running_speed : float, optional
        Speed of the simulated animal in spatial units/second. Default is 15.
    n_runs : int, optional
        Number of runs across the track the simulated animal will perform.
        Default is 15.
    place_field_variance : float, optional
        Spatial extent of place fields (variance of Gaussian). Default is 12.5.
    place_field_means : np.ndarray, shape (n_neurons,), optional
        Location of the center of the Gaussian place fields. Default is
        np.arange(0, 200, 10).
    n_tetrodes : int, optional
        Total number of tetrodes to simulate. Default is 5.
    make_inbound_outbound_neurons : bool, optional
        If True, create neurons with directional place fields (separate neurons
        for inbound vs outbound trajectories). Default is False.
    seed : int | None, optional
        Random seed for reproducibility. If None, uses system randomness.
        Default is 0.

    Returns
    -------
    sim_output : ClusterlessSimOutput
        Dataclass containing:
        - position_time: timestamps for position samples (seconds)
        - position: 2D position array (n_time, n_pos_dims)
        - edges: decoding bin edges (seconds)
        - spike_times: list of spike time arrays per electrode (seconds)
        - spike_waveform_features: list of feature arrays per electrode
        - environment: Environment object
        - bin_widths: bin widths (seconds)

    Notes
    -----
    - All times are in seconds
    - Spike times are strictly increasing per electrode
    - Empty electrodes have shape (0,) for times and (0, n_features) for features
    - No NaN values in spike waveform features

    """
    # Set random seed for reproducibility
    if seed is not None:
        np.random.seed(seed)

    # Generate position trajectory
    n_samples = int(n_runs * sampling_frequency * 2 * track_height / running_speed)
    position_time = simulate_time(n_samples, sampling_frequency)
    position_1d = simulate_position(position_time, track_height, running_speed)
    position = position_1d[:, np.newaxis]  # Shape: (n_time, 1)

    # Generate multiunit data (NaN-padded format) using existing simulator
    multiunits = []
    if not make_inbound_outbound_neurons:
        for place_means in place_field_means.reshape((n_tetrodes, -1)):
            multiunits.append(
                simulate_multiunit_with_place_fields(
                    place_means,
                    position_1d,
                    mark_spacing=10,
                    n_mark_dims=4,
                    place_variance=place_field_variance,
                    sampling_frequency=sampling_frequency,
                )
            )
    else:
        trajectory_direction = get_trajectory_direction(position_1d)
        for direction in np.unique(trajectory_direction):
            is_condition = trajectory_direction == direction
            for place_means in place_field_means.reshape((n_tetrodes, -1)):
                multiunits.append(
                    simulate_multiunit_with_place_fields(
                        place_means,
                        position_1d,
                        mark_spacing=10,
                        n_mark_dims=4,
                        sampling_frequency=sampling_frequency,
                        place_variance=place_field_variance,
                        is_condition=is_condition,
                    )
                )

    # Stack: (n_time, n_features, n_electrodes)
    multiunits = np.stack(multiunits, axis=-1)
    multiunits_spikes = np.any(~np.isnan(multiunits), axis=1)

    # Convert NaN-padded format to per-electrode lists
    n_electrodes = multiunits.shape[2]
    spike_times = []
    spike_waveform_features = []

    for electrode_id in range(n_electrodes):
        spike_indicator = multiunits_spikes[:, electrode_id]
        electrode_spike_times = position_time[spike_indicator].astype(np.float64)
        electrode_features = multiunits[spike_indicator, :, electrode_id].astype(
            np.float64
        )

        # Ensure strictly increasing times (should already be sorted)
        if electrode_spike_times.size > 0:
            assert np.all(np.diff(electrode_spike_times) > 0), (
                "Spike times must be strictly increasing"
            )

        spike_times.append(electrode_spike_times)
        spike_waveform_features.append(electrode_features)

    # Create decoding bin edges (use position sampling rate)
    edges = np.arange(0.0, position_time[-1] + 1e-12, 1.0 / sampling_frequency)
    bin_widths = np.diff(edges)

    # Create environment
    environment = Environment(place_bin_size=1.0)
    environment.fit_place_grid(position)

    return ClusterlessSimOutput(
        position_time=position_time.astype(np.float64),
        position=position.astype(np.float64),
        edges=edges.astype(np.float64),
        spike_times=spike_times,
        spike_waveform_features=spike_waveform_features,
        environment=environment,
        bin_widths=bin_widths.astype(np.float64),
    )


def make_continuous_replay(
    sampling_frequency: int = SAMPLING_FREQUENCY,
    track_height: float = TRACK_HEIGHT,
    running_speed: float = RUNNING_SPEED,
    place_field_means: np.ndarray = PLACE_FIELD_MEANS,
    replay_speedup: float = REPLAY_SPEEDUP,
    n_tetrodes: int = N_TETRODES,
    n_features: int = N_FEATURES,
    mark_spacing: float = MARK_SPACING,
) -> tuple[np.ndarray, np.ndarray]:
    """Creates a simulated continuous replay.

    Parameters
    ----------
    sampling_frequency : int, optional
        Samples per second
    track_height : float, optional
        Height of the simulated track
    running_speed : float, optional
        Simualted speed of the animal
    place_field_means : np.ndarray, optional
        Location of the center of the Gaussian place fields.
    replay_speedup : int, optional
        Number of times faster the replay event is faster than the running speed
    n_tetrodes : int, optional
        Number of simulated tetrodes
    n_features : int, optional
        Number of simulated features
    mark_spacing : float, optional
        Spacing between Gaussian mark features

    Returns
    -------
    replay_time : np.ndarray, shape (n_time,)
        Time in seconds.
    test_multiunits : np.ndarray, shape (n_time, n_features, n_tetrodes)
        Binned clusterless spike times and features. NaN indicates no spike. Non-Nan indicates spike.
    """

    replay_speed = running_speed * replay_speedup
    n_samples = int(0.5 * sampling_frequency * 2 * track_height / replay_speed)
    replay_time = simulate_time(n_samples, sampling_frequency)
    true_replay_position = simulate_position(replay_time, track_height, replay_speed)
    place_field_means = place_field_means.reshape((n_tetrodes, -1))

    min_times_ind = np.argmin(
        np.abs(true_replay_position[:, np.newaxis] - place_field_means.ravel()), axis=0
    )
    tetrode_ind = (
        np.ones_like(place_field_means) * np.arange(5)[:, np.newaxis]
    ).ravel()

    test_multiunits = np.full((replay_time.size, n_features, n_tetrodes), np.nan)
    n_neurons = place_field_means.shape[1]

    mark_centers = np.arange(0, n_neurons * mark_spacing, mark_spacing)
    mark_ind = (np.ones_like(place_field_means) * np.arange(4)).ravel()

    for i in range(n_features):
        test_multiunits[(min_times_ind, i, tetrode_ind)] = mark_centers[mark_ind]

    return replay_time, test_multiunits


def make_hover_replay(
    hover_neuron_ind: int | None = None,
    place_field_means: np.ndarray = PLACE_FIELD_MEANS,
    sampling_frequency: int = SAMPLING_FREQUENCY,
    n_tetrodes: int = N_TETRODES,
    n_features: int = N_FEATURES,
    mark_spacing: float = MARK_SPACING,
) -> tuple[np.ndarray, np.ndarray]:
    """Creates a simulated stationary replay.

    Parameters
    ----------
    hover_neuron_ind : int, optional
        Index of which neuron is the stationary neuron.
    place_field_means : np.ndarray, optional
        Location of the center of the Gaussian place fields.
    sampling_frequency : int, optional
        Samples per second
    n_tetrodes : int, optional
        Number of simulated tetrodes
    n_features : int, optional
        Number of simulated features
    mark_spacing : float, optional
        Spacing between Gaussian mark features

    Returns
    -------
    replay_time : np.ndarray, shape (n_time,)
        Time in seconds.
    test_multiunits : np.ndarray, shape (n_time, n_features, n_tetrodes)
        Binned clusterless spike times and features. NaN indicates no spike. Non-Nan indicates spike.
    """

    place_field_means = place_field_means.reshape((n_tetrodes, -1))

    if hover_neuron_ind is None:
        hover_neuron_ind = place_field_means.size // 2
    tetrode_ind, neuron_ind = np.unravel_index(
        hover_neuron_ind, place_field_means.shape
    )

    N_TIME = 50
    replay_time = np.arange(N_TIME) / sampling_frequency

    spike_time_ind = np.arange(0, N_TIME, 2)
    test_multiunits = np.full((replay_time.size, n_features, n_tetrodes), np.nan)
    n_neurons = place_field_means.shape[1]
    mark_centers = np.arange(0, n_neurons * mark_spacing, mark_spacing)
    test_multiunits[spike_time_ind, :, tetrode_ind] = mark_centers[neuron_ind]

    return replay_time, test_multiunits


def make_fragmented_replay(
    place_field_means: np.ndarray = PLACE_FIELD_MEANS,
    sampling_frequency: int = SAMPLING_FREQUENCY,
    n_tetrodes: int = N_TETRODES,
    n_features: int = N_FEATURES,
    mark_spacing: float = MARK_SPACING,
) -> tuple[np.ndarray, np.ndarray]:
    """Creates a simulated fragmented replay.

    Parameters
    ----------
    place_field_means : np.ndarray, optional
        Location of the center of the Gaussian place fields.
    sampling_frequency : int, optional
        Samples per second
    n_tetrodes : int, optional
        Number of simulated tetrodes
    n_features : int, optional
        Number of simulated features
    mark_spacing : float, optional
        Spacing between Gaussian mark features

    Returns
    -------
    replay_time : np.ndarray, shape (n_time,)
        Time in seconds.
    test_multiunits : np.ndarray, shape (n_time, n_features, n_tetrodes)
        Binned clusterless spike times and features. NaN indicates no spike. Non-Nan indicates spike.
    """

    N_TIME = 10

    place_field_means = place_field_means.reshape((n_tetrodes, -1))
    replay_time = np.arange(N_TIME) / sampling_frequency

    n_total_neurons = place_field_means.size
    neuron_inds = [1, n_total_neurons - 1, 10, n_total_neurons - 5, 8]
    neuron_inds = np.unravel_index(neuron_inds, place_field_means.shape)
    spike_time_ind = [1, 3, 5, 7, 9]
    test_multiunits = np.full((replay_time.size, n_features, n_tetrodes), np.nan)
    n_neurons = place_field_means.shape[1]
    mark_centers = np.arange(0, n_neurons * mark_spacing, mark_spacing)

    for t_ind, tetrode_ind, neuron_ind in zip(
        spike_time_ind, *neuron_inds, strict=False
    ):
        test_multiunits[t_ind, :, tetrode_ind] = mark_centers[neuron_ind]

    return replay_time, test_multiunits


def make_hover_continuous_hover_replay(
    sampling_frequency: int = SAMPLING_FREQUENCY,
    place_field_means: np.ndarray = PLACE_FIELD_MEANS,
) -> tuple[np.ndarray, np.ndarray]:
    """Make a simulated replay that first is stationary, then is continuous, then is stationary again.

    Parameters
    ----------
    sampling_frequency : int, optional
        Samples per second
    place_field_means : np.ndarray, optional
        Location of the center of the Gaussian place fields.

    Returns
    -------
    replay_time : np.ndarray, shape (n_time,)
        Time in seconds.
    test_multiunits : np.ndarray, shape (n_time, n_features, n_tetrodes)
        Binned clusterless spike times and features. NaN indicates no spike. Non-Nan indicates spike.
    """
    _, test_multiunits1 = make_hover_replay(hover_neuron_ind=0)
    _, test_multiunits2 = make_continuous_replay()
    n_total_neurons = place_field_means.size
    _, test_multiunits3 = make_hover_replay(hover_neuron_ind=n_total_neurons - 1)

    test_multiunits = np.concatenate(
        (test_multiunits1, test_multiunits2, test_multiunits3)
    )
    replay_time = np.arange(test_multiunits.shape[0]) / sampling_frequency

    return replay_time, test_multiunits


def make_fragmented_hover_fragmented_replay(
    sampling_frequency: int = SAMPLING_FREQUENCY,
) -> tuple[np.ndarray, np.ndarray]:
    """Makes a simulated replay that first is fragmented, then is stationary, then is fragmented again.

    Parameters
    ----------
    sampling_frequency : int, optional
        Samples per second

    Returns
    -------
    replay_time : np.ndarray, shape (n_time,)
        Time in seconds.
    test_multiunits : np.ndarray, shape (n_time, n_features, n_tetrodes)
        Binned clusterless spike times and features. NaN indicates no spike. Non-Nan indicates spike.
    """
    _, test_multiunits1 = make_fragmented_replay()
    _, test_multiunits2 = make_hover_replay(hover_neuron_ind=6)
    _, test_multiunits3 = make_fragmented_replay()

    test_multiunits = np.concatenate(
        (test_multiunits1, test_multiunits2, test_multiunits3)
    )
    replay_time = np.arange(test_multiunits.shape[0]) / sampling_frequency

    return replay_time, test_multiunits


def make_fragmented_continuous_fragmented_replay(
    sampling_frequency: int = SAMPLING_FREQUENCY,
) -> tuple[np.ndarray, np.ndarray]:
    """Makes a simulated replay that first is fragmented, then is continuous, then is fragmented again.

    Parameters
    ----------
    sampling_frequency : int, optional
        Samples per second

    Returns
    -------
    replay_time : np.ndarray, shape (n_time,)
        Time in seconds.
    test_multiunits : np.ndarray, shape (n_time, n_features, n_tetrodes)
        Binned clusterless spike times and features. NaN indicates no spike. Non-Nan indicates spike.

    """
    _, test_multiunits1 = make_fragmented_replay()
    _, test_multiunits2 = make_continuous_replay()
    _, test_multiunits3 = make_fragmented_replay()

    test_multiunits = np.concatenate(
        (test_multiunits1, test_multiunits2, test_multiunits3)
    )
    replay_time = np.arange(test_multiunits.shape[0]) / sampling_frequency

    return replay_time, test_multiunits
