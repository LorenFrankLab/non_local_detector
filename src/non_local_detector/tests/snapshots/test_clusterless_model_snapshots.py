"""Snapshot tests for clusterless decoder models.

These tests verify consistent behavior of clusterless decoders that work
with continuous spike features (waveform features) rather than sorted units.
"""

import numpy as np
import pytest
from syrupy.assertion import SnapshotAssertion

from non_local_detector import (
    ClusterlessDecoder,
    ContFragClusterlessClassifier,
    NonLocalClusterlessDetector,
)
from non_local_detector.simulate.clusterless_simulation import make_simulated_run_data


@pytest.fixture
def clusterless_simulated_data():
    """Generate simulated clusterless data with fixed random seed."""
    n_tetrodes = 4
    place_field_means = np.arange(0, 160, 10)  # 16 place fields evenly divisible by 4

    # Use new API that returns ClusterlessSimOutput directly
    sim = make_simulated_run_data(
        sampling_frequency=500,
        n_runs=5,
        n_tetrodes=n_tetrodes,
        place_field_means=place_field_means,
        seed=42,  # Fixed seed for reproducibility
    )

    return {
        "time": sim.position_time,
        "position": sim.position,  # Already 2D from new API
        "spike_times": sim.spike_times,
        "spike_waveform_features": sim.spike_waveform_features,
        "sampling_frequency": 500,
    }


def serialize_xarray_summary(data_array):
    """Serialize xarray DataArray to summary statistics."""
    arr = np.asarray(data_array)
    return {
        "shape": arr.shape,
        "dtype": str(arr.dtype),
        "mean": float(np.mean(arr)),
        "std": float(np.std(arr)),
        "min": float(np.min(arr)),
        "max": float(np.max(arr)),
        "sum": float(np.sum(arr)),
        "first_5": arr.ravel()[:5].tolist() if arr.size >= 5 else arr.ravel().tolist(),
        "last_5": arr.ravel()[-5:].tolist() if arr.size >= 5 else arr.ravel().tolist(),
    }


def serialize_state_probabilities(state_probs):
    """Serialize state probabilities for snapshot comparison."""
    return {
        "shape": state_probs.shape,
        "states": list(state_probs.states.values),
        "mean_per_state": {
            str(state): float(np.mean(state_probs.sel(states=state)))
            for state in state_probs.states.values
        },
        "max_per_state": {
            str(state): float(np.max(state_probs.sel(states=state)))
            for state in state_probs.states.values
        },
        "min_per_state": {
            str(state): float(np.min(state_probs.sel(states=state)))
            for state in state_probs.states.values
        },
    }


@pytest.mark.snapshot
@pytest.mark.slow
@pytest.mark.skip(
    reason="Skipped due to known issues in clusterless_kde likelihood code. "
    "Will be re-enabled once clusterless_kde is fixed."
)
@pytest.mark.parametrize(
    "algorithm,algorithm_params",
    [
        (
            "clusterless_kde",
            {"position_std": 6.0, "waveform_std": 24.0, "block_size": 100},
        ),
    ],
)
def test_nonlocal_clusterless_detector_snapshot(
    clusterless_simulated_data: dict,
    algorithm: str,
    algorithm_params: dict,
    snapshot: SnapshotAssertion,
):
    """Snapshot test for NonLocalClusterlessDetector with KDE algorithm.

    Tests KDE-based clusterless likelihood model.
    Note: GMM algorithm excluded due to different parameter requirements.
    Note: Currently skipped due to data sparsity issues with small test subsets.
    """
    # Use a subset of data for faster testing
    n_samples = min(5000, len(clusterless_simulated_data["time"]))
    is_training = np.ones(n_samples, dtype=bool)
    is_training[2000:2500] = False  # Leave some data for testing

    detector = NonLocalClusterlessDetector(
        clusterless_algorithm=algorithm,
        clusterless_algorithm_params=algorithm_params,
    ).fit(
        clusterless_simulated_data["time"][:n_samples],
        clusterless_simulated_data["position"][:n_samples],
        [
            st[st < clusterless_simulated_data["time"][n_samples]]
            for st in clusterless_simulated_data["spike_times"]
        ],
        [
            swf[
                clusterless_simulated_data["spike_times"][i]
                < clusterless_simulated_data["time"][n_samples]
            ]
            for i, swf in enumerate(
                clusterless_simulated_data["spike_waveform_features"]
            )
        ],
        is_training=is_training,
    )

    results = detector.predict(
        spike_times=[
            st[
                (st >= clusterless_simulated_data["time"][2000])
                & (st < clusterless_simulated_data["time"][2500])
            ]
            for st in clusterless_simulated_data["spike_times"]
        ],
        spike_waveform_features=[
            swf[
                (
                    clusterless_simulated_data["spike_times"][i]
                    >= clusterless_simulated_data["time"][2000]
                )
                & (
                    clusterless_simulated_data["spike_times"][i]
                    < clusterless_simulated_data["time"][2500]
                )
            ]
            for i, swf in enumerate(
                clusterless_simulated_data["spike_waveform_features"]
            )
        ],
        time=clusterless_simulated_data["time"][2000:2500],
        position=clusterless_simulated_data["position"][2000:2500],
        position_time=clusterless_simulated_data["time"][2000:2500],
    )

    # Snapshot state probabilities
    state_probs_summary = serialize_state_probabilities(
        results.acausal_state_probabilities
    )
    assert state_probs_summary == snapshot(name="state_probabilities")

    # Snapshot posterior summary
    posterior_summary = serialize_xarray_summary(results.acausal_posterior)
    assert posterior_summary == snapshot(name="posterior_summary")


@pytest.mark.snapshot
def test_contfrag_clusterless_classifier_snapshot(
    clusterless_simulated_data: dict, snapshot: SnapshotAssertion
):
    """Snapshot test for ContFragClusterlessClassifier."""
    # Use a subset of data for faster testing
    n_samples = min(5000, len(clusterless_simulated_data["time"]))
    is_training = np.ones(n_samples, dtype=bool)
    is_training[2000:2500] = False

    classifier = ContFragClusterlessClassifier(
        clusterless_algorithm="clusterless_kde",
        clusterless_algorithm_params={
            "position_std": 6.0,
            "waveform_std": 24.0,
            "block_size": 100,
        },
    ).fit(
        position_time=clusterless_simulated_data["time"][:n_samples],
        position=clusterless_simulated_data["position"][:n_samples],
        spike_times=[
            st[st < clusterless_simulated_data["time"][n_samples]]
            for st in clusterless_simulated_data["spike_times"]
        ],
        spike_waveform_features=[
            swf[
                clusterless_simulated_data["spike_times"][i]
                < clusterless_simulated_data["time"][n_samples]
            ]
            for i, swf in enumerate(
                clusterless_simulated_data["spike_waveform_features"]
            )
        ],
        is_training=is_training,
    )

    results = classifier.predict(
        spike_times=[
            st[
                (st >= clusterless_simulated_data["time"][2000])
                & (st < clusterless_simulated_data["time"][2500])
            ]
            for st in clusterless_simulated_data["spike_times"]
        ],
        spike_waveform_features=[
            swf[
                (
                    clusterless_simulated_data["spike_times"][i]
                    >= clusterless_simulated_data["time"][2000]
                )
                & (
                    clusterless_simulated_data["spike_times"][i]
                    < clusterless_simulated_data["time"][2500]
                )
            ]
            for i, swf in enumerate(
                clusterless_simulated_data["spike_waveform_features"]
            )
        ],
        time=clusterless_simulated_data["time"][2000:2500],
        position=clusterless_simulated_data["position"][2000:2500],
        position_time=clusterless_simulated_data["time"][2000:2500],
    )

    # Snapshot state probabilities
    state_probs_summary = serialize_state_probabilities(
        results.acausal_state_probabilities
    )
    assert state_probs_summary == snapshot(name="state_probabilities")

    # Snapshot posterior summary
    posterior_summary = serialize_xarray_summary(results.acausal_posterior)
    assert posterior_summary == snapshot(name="posterior_summary")


@pytest.mark.snapshot
def test_clusterless_decoder_snapshot(
    clusterless_simulated_data: dict, snapshot: SnapshotAssertion
):
    """Snapshot test for basic ClusterlessDecoder."""
    # Use a subset of data for faster testing
    n_samples = min(5000, len(clusterless_simulated_data["time"]))
    is_training = np.ones(n_samples, dtype=bool)
    is_training[2000:2500] = False

    decoder = ClusterlessDecoder(
        clusterless_algorithm="clusterless_kde",
        clusterless_algorithm_params={
            "position_std": 6.0,
            "waveform_std": 24.0,
            "block_size": 100,
        },
    ).fit(
        position_time=clusterless_simulated_data["time"][:n_samples],
        position=clusterless_simulated_data["position"][:n_samples],
        spike_times=[
            st[st < clusterless_simulated_data["time"][n_samples]]
            for st in clusterless_simulated_data["spike_times"]
        ],
        spike_waveform_features=[
            swf[
                clusterless_simulated_data["spike_times"][i]
                < clusterless_simulated_data["time"][n_samples]
            ]
            for i, swf in enumerate(
                clusterless_simulated_data["spike_waveform_features"]
            )
        ],
        is_training=is_training,
    )

    results = decoder.predict(
        spike_times=[
            st[
                (st >= clusterless_simulated_data["time"][2000])
                & (st < clusterless_simulated_data["time"][2500])
            ]
            for st in clusterless_simulated_data["spike_times"]
        ],
        spike_waveform_features=[
            swf[
                (
                    clusterless_simulated_data["spike_times"][i]
                    >= clusterless_simulated_data["time"][2000]
                )
                & (
                    clusterless_simulated_data["spike_times"][i]
                    < clusterless_simulated_data["time"][2500]
                )
            ]
            for i, swf in enumerate(
                clusterless_simulated_data["spike_waveform_features"]
            )
        ],
        time=clusterless_simulated_data["time"][2000:2500],
        position=clusterless_simulated_data["position"][2000:2500],
        position_time=clusterless_simulated_data["time"][2000:2500],
    )

    # Snapshot state probabilities
    state_probs_summary = {
        "shape": results.acausal_state_probabilities.shape,
        "mean": float(np.mean(results.acausal_state_probabilities)),
        "min": float(np.min(results.acausal_state_probabilities)),
        "max": float(np.max(results.acausal_state_probabilities)),
    }
    assert state_probs_summary == snapshot(name="state_probabilities")

    # Snapshot posterior summary
    posterior_summary = serialize_xarray_summary(results.acausal_posterior)
    assert posterior_summary == snapshot(name="posterior_summary")


@pytest.mark.snapshot
def test_clusterless_encoding_model_snapshot(
    clusterless_simulated_data: dict, snapshot: SnapshotAssertion
):
    """Snapshot test for clusterless encoding model properties."""
    n_samples = min(5000, len(clusterless_simulated_data["time"]))

    decoder = ClusterlessDecoder(
        clusterless_algorithm="clusterless_kde",
        clusterless_algorithm_params={
            "position_std": 6.0,
            "waveform_std": 24.0,
            "block_size": 100,
        },
    ).fit(
        position_time=clusterless_simulated_data["time"][:n_samples],
        position=clusterless_simulated_data["position"][:n_samples],
        spike_times=[
            st[st < clusterless_simulated_data["time"][n_samples]]
            for st in clusterless_simulated_data["spike_times"]
        ],
        spike_waveform_features=[
            swf[
                clusterless_simulated_data["spike_times"][i]
                < clusterless_simulated_data["time"][n_samples]
            ]
            for i, swf in enumerate(
                clusterless_simulated_data["spike_waveform_features"]
            )
        ],
    )

    # Get encoding model for first environment
    enc_model = decoder.encoding_model_[("", 0)]

    # Snapshot occupancy
    occupancy_summary = {
        "shape": enc_model["occupancy"].shape,
        "mean": float(np.mean(enc_model["occupancy"])),
        "min": float(np.min(enc_model["occupancy"])),
        "max": float(np.max(enc_model["occupancy"])),
        "sum": float(np.sum(enc_model["occupancy"])),
    }
    assert occupancy_summary == snapshot(name="occupancy_summary")

    # Snapshot mean rates
    mean_rates_summary = {
        "n_electrodes": len(enc_model["mean_rates"]),
        "rates": [float(rate) for rate in enc_model["mean_rates"]],
    }
    assert mean_rates_summary == snapshot(name="mean_rates_summary")
