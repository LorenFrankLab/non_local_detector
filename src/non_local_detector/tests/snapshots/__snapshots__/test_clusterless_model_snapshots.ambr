# serializer version: 1
# name: test_clusterless_decoder_snapshot[posterior_summary]
  dict({
    'dtype': 'float32',
    'first_5': list([
      1.9468870959826745e-08,
      5.433183503100736e-08,
      1.7823427356233879e-07,
      5.957490998298454e-07,
      1.8952069922306691e-06,
    ]),
    'last_5': list([
      3.466594781502743e-19,
      7.31018281404324e-20,
      1.6921430227018417e-20,
      4.403664011987968e-21,
      1.2738280530212939e-21,
    ]),
    'max': 0.18298926949501038,
    'mean': 0.011904762126505375,
    'min': 0.0,
    'shape': tuple(
      500,
      84,
    ),
    'std': 0.031634386628866196,
    'sum': 500.0,
  })
# ---
# name: test_clusterless_decoder_snapshot[state_probabilities]
  dict({
    'max': 1.000000238418579,
    'mean': 1.0,
    'min': 0.9999997615814209,
    'shape': tuple(
      500,
    ),
  })
# ---
# name: test_clusterless_encoding_model_snapshot[mean_rates_summary]
  dict({
    'n_electrodes': 4,
    'rates': list([
      0.058400001376867294,
      0.03139999881386757,
      0.032999999821186066,
      0.03180000185966492,
    ]),
  })
# ---
# name: test_clusterless_encoding_model_snapshot[occupancy_summary]
  dict({
    'max': 0.011862837709486485,
    'mean': 0.005550175439566374,
    'min': 0.0042543173767626286,
    'shape': tuple(
      84,
    ),
    'sum': 0.46621471643447876,
  })
# ---
# name: test_contfrag_clusterless_classifier_snapshot[posterior_summary]
  dict({
    'dtype': 'float32',
    'first_5': list([
      9.055584814632311e-05,
      9.05100314412266e-05,
      9.03815234778449e-05,
      9.03200707398355e-05,
      9.087444777833298e-05,
    ]),
    'last_5': list([
      0.0030791168101131916,
      0.0031272100750356913,
      0.003167081857100129,
      0.0031980357598513365,
      0.0032208815682679415,
    ]),
    'max': 0.18263192474842072,
    'mean': 0.0059523810632526875,
    'min': 8.516813052501543e-17,
    'shape': tuple(
      500,
      168,
    ),
    'std': 0.02291184291243553,
    'sum': 500.0,
  })
# ---
# name: test_contfrag_clusterless_classifier_snapshot[state_probabilities]
  dict({
    'max_per_state': dict({
      'Continuous': 0.9999198913574219,
      'Fragmented': 0.2347736805677414,
    }),
    'mean_per_state': dict({
      'Continuous': 0.9897478222846985,
      'Fragmented': 0.010252131149172783,
    }),
    'min_per_state': dict({
      'Continuous': 0.7652262449264526,
      'Fragmented': 8.02369904704392e-05,
    }),
    'shape': tuple(
      500,
      2,
    ),
    'states': list([
      np.str_('Continuous'),
      np.str_('Fragmented'),
    ]),
  })
# ---
